src/utilities/createRouterHooks.ts

```diff
@@ -0,0 +1,77 @@
+import { InjectionKey } from 'vue'
+import { RouteMiddleware } from '@/types'
+import { asArray } from '@/utilities/array'
+import { RouterRoute } from '@/utilities/createRouterRoute'
+
+export type RouteHookRemove = () => void
+export type AddRouteHook = (hook: RouteMiddleware) => RouteHookRemove
+export type RouteHookType = 'before' | 'after'
+export type RouteHookLifeCycle = 'onBeforeRouteEnter' | 'onBeforeRouteLeave' | 'onBeforeRouteUpdate'
+
+type AddRouteHookForLifeCycle = (type: RouteHookType, hook: RouteMiddleware) => RouteHookRemove
+type RouteHooks = Record<RouteHookType, Set<RouteMiddleware>>
+
+export const addRouteHookInjectionKey: InjectionKey<AddRouteHookForLifeCycle> = Symbol()
+
+export type RouterHooks = {
+  onBeforeRouteEnter: AddRouteHook,
+  onBeforeRouteUpdate: AddRouteHook,
+  onBeforeRouteLeave: AddRouteHook,
+  addRouteHook: AddRouteHookForLifeCycle,
+  hooks: RouteHooks,
+}
+
+type GlobalHookCondition = (to: RouterRoute, from: RouterRoute | null) => boolean
+
+const isGlobalEnter: GlobalHookCondition = (to, from) => {
+  return to.matches[1] !== from?.matches[1]
+}
+
+const isGlobalUpdate: GlobalHookCondition = (to, from) => {
+  return to.matches.some((route, depth) => route === from?.matches[depth])
+}
+
+const isGlobalLeave: GlobalHookCondition = (to, from) => {
+  return to.matches[1] !== from?.matches[1]
+}
```

Comment:
These are the lifecycle conditions for global hooks. Currently enter and leave are the same. But seemed worth separating out for consistency and easy naming.